src/scratchOrgSettingsGenerator.ts

```diff
@@ -0,0 +1,421 @@
+  fullName: string;
+  isActive: boolean;
+  values: [
+    {
+      fullName: string;
+      default?: boolean;
+    }
+  ];
+}
+
+/**
+ * Helper class for dealing with the settings that are defined in a scratch definition file.  This class knows how to extract the
+ * settings from the definition, how to expand them into a MD directory and how to generate a package.xml.
+ */
+export default class SettingsGenerator {
+  private settingData: Nullable<Record<string, unknown>>;
+  private objectSettingsData: Nullable<{ [objectName: string]: ObjectSetting }>;
+  private currentApiVersion = ensureString(getApiVersion());
+  private logger: Logger;
+
+  public constructor() {
+    this.logger = Logger.childFromRoot('SettingsGenerator');
+  }
+
+  /** extract the settings from the scratch def file, if they are present. */
+  public async extract(scratchDef: ScratchDefinition, apiVersion?: string): Promise<void> {
+    this.logger.debug('extracting settings from scratch definition file');
+    if (!apiVersion) {
+      apiVersion = this.currentApiVersion;
+    }
+    if (parseFloat(apiVersion) >= 47.0 && this.orgPreferenceSettingsMigrationRequired(scratchDef)) {
+      this.extractAndMigrateSettings(scratchDef);
+    } else {
+      this.settingData = getObject(scratchDef, 'settings');
+      this.objectSettingsData = getObject(scratchDef, 'objectSettings');
+    }
+
+    this.logger.debug('settings are', this.settingData);
+    // TODO, this is where we will validate the settings.
+    // See W-5068155
+    // if (this.hasSettings()) {  }
+  }
+
+  /** True if we are currently tracking setting or object setting data. */
+  public hasSettings(): boolean {
+    return !(isEmpty(this.settingData) && isEmpty(this.objectSettingsData));
+  }
+
+  /** Check to see if the scratchDef contains orgPreferenceSettings
+   *  orgPreferenceSettings are no longer supported after api version 46.0
+   */
+  public orgPreferenceSettingsMigrationRequired(scratchDef: ScratchDefinition): boolean {
+    return !(!scratchDef || !scratchDef.settings || !scratchDef.settings.orgPreferenceSettings);
+  }
+
+  /** This will copy all of the settings in the scratchOrgInfo orgPreferences mapping into the settings structure.
+   *  It will also spit out a warning about the pending deprecation og the orgPreferences structure.
+   *  This returns a failure message in the promise upon critical error for api versions after 46.0.
+   *  For api versions less than 47.0 it will return a warning.
+   */
+  public migrate(scratchDef: ScratchDefinition, apiVersion?: string): Optional<string[]> {
+    const warnings: string[] = [];
+    // Make sure we have old style preferences
+    if (!scratchDef.orgPreferences) {
+      return warnings;
+    }
+
+    if (!apiVersion) {
+      apiVersion = this.currentApiVersion;
+    }
+
+    // First, let's map the old style tooling preferences into MD-API preferences
+    this.settingData = {};
+
+    function lhccmdt(mdt: string): string {
```

Comment:
Would prefer the use of npm module [change-case](https://www.npmjs.com/package/change-case) for these types of operations.